FISH-1342 - Support for loading certificates from multiple keystores (#182)
payara/Payara-Community-Documentation: docs/modules/ROOT/pages/documentation/payara-server/server-configuration/ssl-certificates.adoc

The information can also be entered in the Web Administration console, by navigating to the page Configurations -> server-config -> HTTP Service -> HTTP Listeners > http-listener-2.

[[loading-certificates-from-multiple-keystores]]
== Loading certificates from multiple keystores

IMPORTANT: All additional keystores must have the same password as your domain master password.

To add additional keystores to the Payara Server you will need to have your keystore and truststore files in a local directory. You can then configure this either in the admin console or using asadmin commands. If you want to load more than one additional keystore use ':' as a delimiter between file paths.

The new JVM properties used to add additional keystores are:

> * -Dfish.payara.ssl.additionalKeyStores

Reviewer: Should probably be formatted using code tags

* -Dfish.payara.ssl.additionalTrustStores

=== Using the Admin Console
To configure the additional keystore locations in the admin console, head to the *Configurations -> <instance configuration> -> JVM Settings* and on the *JVM Options* tab click *Add JVM Option*.

> In the new blank JVM property, you can add the new JVM property and the relative paths to your keystores and truststores respectively.

Reviewer: ```suggestion
In the new blank JVM property, you can add the new JVM property and the relative paths to your keystores or truststores respectively.
```

image::ssl/add-additional-keystores-admin-console.png[Configure additional keystores and truststores in Admin Console]

=== Using Asadmin Commands

> JVM options can be configured using the 'create-jvm-options' asadmin command, you can configure your additional keystore and truststore files like in the following examples. If you are loading in multiple additional keystores via asadmin commands, you will need to use '\:' as the delimeter to avoid creating multiple JVM options.

Reviewer: ```suggestion
JVM options can be configured using the 'create-jvm-options' asadmin command, you can configure your additional keystore and truststore files like in the following examples. If you are loading in multiple additional keystores via asadmin commands, you will need to use '\:' as the delimiter to avoid creating multiple JVM options.
```

The information can also be entered in the Web Administration console, by navigating to the page Configurations -> server-config -> HTTP Service -> HTTP Listeners > http-listener-2.

[[loading-certificates-from-multiple-keystores]]
== Loading certificates from multiple keystores

> _Since Payara Server 5.2021.5_

Reviewer: I don't think we put these in any more.

> IMPORTANT: All additional keystores must have the same password as your domain master password. If you want to load more than one additional keystore use ':' as a delimiter between file paths.

Reviewer: I'd move the bit about loading multiple keystores out from here - having two separate points in the same caption strikes me as bad practice.

To add additional keystores to the Payara Server you will need to have your keystore and truststore files in a local directory. You can then configure this either in the admin console or using asadmin commands.

The new JVM properties used to add additional keystores are:

* -Dfish.payara.ssl.additionalKeyStores
* -Dfish.payara.ssl.additionalTrustStores

=== Using the Admin Console
To configure the additional keystore locations in the admin console, head to the *Configurations -> <instance configuration> -> JVM Settings* and on the *JVM Options* tab click *Add JVM Option*.

> In the new blank JVM property, you can add the new JVM properties and the relative paths to your keystores and truststores respectively.

Reviewer: `new blank JVM propert**y**` ... `new JVM propert**ies**`
The grammar pedant in me cries

image::ssl/add-additional-keystores-admin-console.png[Configure additional keystores and truststores in Admin Console]

=== Using Asadmin Commands
JVM options can be configured using the 'create-jvm-options' asadmin command, you can configure your additional keystore and truststore files like in the following examples:

*Additional Keystores*
[source, shell]
----
asadmin create-jvm-options "-Dfish.payara.ssl.additionalKeyStores=/path/to/keystore.jks:/path2/to/keystore2.jks"
----
*Additional Truststores*
[source, shell]
----

> asadmin create-jvm-options "-Dfish.payara.ssl.additionalTrustStores=/path/to/truststore.jks:/path2/to/truststore2.jks"

Reviewer: I'm not sure this is true...
The colon is used to differentiate different JVM options, so surely what you'd get here is two separate options:
* `-Dfish.payara.ssl.additionalTrustStores=/path/to/truststore.jks`
* `/path/to/truststore2.jks`

----

> NOTE: If you load multiple keystores with the same alias, the server will use the keystore with that alias declared first in the JVM options.

Reviewer: Is this true? Would it not default to the default keystore first?
Surely the order would be: default, additional1, additional2, ..., additionaln

The information can also be entered in the Web Administration console, by navigating to the page Configurations -> server-config -> HTTP Service -> HTTP Listeners > http-listener-2.

[[loading-certificates-from-multiple-keystores]]
== Loading certificates from multiple keystores

IMPORTANT: All additional keystores must have the same password as your domain master password.

> To add additional keystores to the Payara Server you will need to have your keystore and truststore files in a local directory. You can then configure this either in the admin console or using asadmin commands. If you want to load more than one additional keystore use ':' as a delimiter between file paths.

Reviewer 2: ```suggestion
To add additional keystores to Payara Server you will need to have your keystore and truststore files in a local directory. You can then configure this using the admin console or asadmin commands. If you want to load more than one additional keystore use ':' as a delimiter between file paths.
```

NOTE: When specifying the key or trust store for a specific listener, this is deemed as an absolute and any additional key or trust stores are not considered.

The new JVM properties used to add additional keystores are:
----
-Dfish.payara.ssl.additionalKeyStores
----
----
-Dfish.payara.ssl.additionalTrustStores
----
=== Using the Admin Console
To configure the additional keystore locations in the admin console, head to the *Configurations -> <instance configuration> -> JVM Settings* and on the *JVM Options* tab click *Add JVM Option*.

> In the new blank JVM property, you can add the new JVM property and the relative paths to your keystores or truststores respectively.

Reviewer 2: ```suggestion
You can add the new JVM property and the relative paths to your keystores or truststores here.
```

image::ssl/add-additional-keystores-admin-console.png[Configure additional keystores and truststores in Admin Console]

=== Using Asadmin Commands

> JVM options can be configured using the 'create-jvm-options' asadmin command, you can configure your additional keystore and truststore files like in the following examples. If you are loading in multiple additional keystores via asadmin commands, you will need to use '\:' as the delimiter to avoid creating multiple JVM options.

Reviewer 2: ```suggestion
JVM options can be configured using the 'create-jvm-options' asadmin command, you can configure your additional keystore and truststore files using this command. If you are loading in multiple additional keystores via asadmin commands, you will need to use '\:' as the delimiter to avoid creating multiple JVM options.
```
